### DISCUSSION TITLE
Send structs using ports

### DISCUSSION THREAD

**Discussion by sheershak-agarwal** [nasa/fprime#1299] [Q&A]
Hello,

I have a really big struct in Gps and I want to send the whole struct to another component through a port for logging. Instead of creating individual argument, I want to send the pointer or better a reference to the whole struct.

So my port xml has:
```
<interface name="GpsToTlm" namespace="App">
<args>
   <arg name="data" type="I8" pass_by="pointer"> </arg>
   <arg name="size" type="U32"> </arg>
</args>
</interface>
```
Sending using this->GpsToTlmOut_out((I8 *)&gps_data, sizeof(gps_data));
where gps_data is init as:
struct GPS_DATA {
 F64 lan,
 F64 lon
};

how do I send it as a reference? 

I also saw all the serializable xmls. Should I create one of those? How can I use it inside my cpp file? 
If I want to write the serializable object to a binary file, do I have to use serialize functions? because write now I use ```gps_file.write((char *)&gps_data, size);``` within the component.
Can I also send serializables through events? if yes, what will be their format_string?

Thanks

**Comment by timcanham** [comment]
You can define the struct one of two ways:

1) define an autocoded serializable type in XML or FPP (assuming you are still using XML):

Here is an example: https://github.com/nasa/fprime/blob/v2.1.0/Ref/SignalGen/SignalPairSerializableAi.xml

2) Define a hand coded serializable:

Here is an example: https://github.com/nasa/fprime/blob/v2.1.0/Autocoders/Python/test/serialize_user/UserSerializer.hpp

With 2), you have to implement the functions to serialize and deserialize any internal data.

Once you define them, you can make them arguments in ports and pass by value:

https://github.com/nasa/fprime/blob/v2.1.0/Autocoders/Python/test/serialize_user/ExamplePortAi.xml

Note that you can use 1) for telemetry or events, but not 2).

**Reply by sheershak-agarwal** [reply]
Can I create arrays in Serializable xml like this? It is throwing an error when I build
```
<member name="pos" type="F64" array_size="3"/>
```

**Reply by timcanham** [reply]
I haven't experimented with that, so give it a try! @LeStarch?

**Reply by sheershak-agarwal** [reply]
It definitely fails for me after adding the array_size. I will proceed without array for now. Thanks

**Reply by timcanham** [reply]
What do the 3 F64 values represent?

**Reply by sheershak-agarwal** [reply]
Can I initialize a serializable object with all zeros?

**Reply by sheershak-agarwal** [reply]
the 3 F64 vales are lat, lon, and alt

**Reply by timcanham** [reply]
If they are functionally different, make each of them a member. It makes it easier to track which one is which. The generated code has a constructor and `set()` methods to set them to whatever you want.

**Reply by sheershak-agarwal** [reply]
btw I'm using v2.1.0

**Reply by sheershak-agarwal** [reply]
and  `<member` name="temperatures" type="F64" size="8"/>` works for me.

**Comment by sheershak-agarwal** [comment]
what is the best way to write the serializable object to a file? and then deserialize the file at ground probably by gds or custom python?
I am currently using the toString method. Is there a better way? because it take a 32 bytes object to 212 bytes.

**Reply by sheershak-agarwal** [reply]
```CPP

// string 
App::OBC_DATA data
Fw::String str;
data.toString(str);
printf("Received OBC_DATA %s --- %u\n", str.toChar(), str.length());

// ComBuffer
Fw::ComBuffer buf;
buf.resetSer();
Fw::SerializeStatus serStat = buf.serialize(data);
FW_ASSERT(Fw::FW_SERIALIZE_OK == serStat,(NATIVE_INT_TYPE)serStat);
printf("Received OBC_DATA %s --- %u\n", buf.getBuffAddr(), buf.getBuffLength());
```
So I can serialize data and put it in text file. How can I deserialize it in python?
